clarity/contracts/vaults-v2/arkadiko-vaults-operations-v1-1.clar

```diff
@@ -0,0 +1,172 @@
+  )
+    (asserts! (not (is-eq (get status vault) STATUS_ACTIVE)) (err ERR_WRONG_STATUS))
+    (asserts! (get valid coll-to-debt) (err ERR_INVALID_RATIO))
+    (asserts! (< (+ total-debt debt) (get max-debt collateral-info)) (err ERR_MAX_DEBT_REACHED))
+
+    ;; Save vault data
+    (try! (as-contract (contract-call? .arkadiko-vaults-data-v1-1 set-vault owner (contract-of token) STATUS_ACTIVE collateral debt)))
+    (try! (as-contract (contract-call? .arkadiko-vaults-sorted-v1-1 insert owner (contract-of token) nicr prev-owner-hint next-owner-hint)))
+
+    ;; Deposit collateral
+    (try! (contract-call? .arkadiko-vaults-pool-active-v1-1 deposit token owner collateral))
+
+    ;; Mint USDA
+    (try! (as-contract (contract-call? .arkadiko-dao mint-token .usda-token debt owner)))
+
+    (ok true)
+  )
+)
+
+;; Update vault
+;; Can change collateral and/or debt
+;; Need hints to know where to insert vault in sorted list
+(define-public (update-vault 
+  (oracle <oracle-trait>) 
+  (token <ft-trait>) 
+  (collateral uint) 
+  (debt uint) 
+  (prev-owner-hint (optional principal)) 
+  (next-owner-hint (optional principal))
+)
+  (let (
+    (owner tx-sender)
+    (stability-fee (unwrap-panic (get-stability-fee owner (contract-of token))))
+    (nicr (/ (* collateral u100000000) debt))
+    (collateral-info (unwrap! (contract-call? .arkadiko-vaults-tokens-v1-1 get-token (contract-of token)) (err ERR_UNKNOWN_TOKEN)))
+    (vault (contract-call? .arkadiko-vaults-data-v1-1 get-vault owner (contract-of token)))
+    (total-debt (get total (contract-call? .arkadiko-vaults-data-v1-1 get-total-debt (contract-of token))))
+    (coll-to-debt (try! (get-collateral-to-debt oracle owner (contract-of token) collateral debt)))
+  )
+    (asserts! (is-eq (get status vault) STATUS_ACTIVE) (err ERR_WRONG_STATUS))
+    (asserts! (get valid coll-to-debt) (err ERR_INVALID_RATIO))
+    (asserts! (< (+ (- total-debt (get debt vault)) debt) (get max-debt collateral-info)) (err ERR_MAX_DEBT_REACHED))
+
+    ;; Update vault data
+    (try! (as-contract (contract-call? .arkadiko-vaults-data-v1-1 set-vault owner (contract-of token) STATUS_ACTIVE collateral debt)))
+    (try! (as-contract (contract-call? .arkadiko-vaults-sorted-v1-1 reinsert owner (contract-of token) nicr prev-owner-hint next-owner-hint)))
+
+    ;; Mint or burn USDA
+    (if (is-eq debt (get debt vault))
+      false
+      (if (> debt (get debt vault))
+        (try! (as-contract (contract-call? .arkadiko-dao mint-token .usda-token (- debt (get debt vault)) owner)))
+        (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token (- (get debt vault) debt) owner)))
+      )
+    )
+
+    ;; Deposit or withdraw collateral
+    (if (is-eq collateral (get collateral vault))
+      false
+      (if (> collateral (get collateral vault))
+        (try! (contract-call? .arkadiko-vaults-pool-active-v1-1 deposit token owner (- collateral (get collateral vault))))
```

Comment:
Q: I think we have to make sure the `token` is whitelisted, no? If we allow any SIP10, the `transfer` function can do malicious things

Reply:
A: At the beginning of this method we get the token info from the collateral-types, which throws ERR_UNKNOWN_TOKEN if token not found.